File: python_rpn_solver/solver/parse/tokenizer.py
from ..token.token import Token
from ..token.literal_token import LiteralToken
from ..token.operator_token import Operator, OperatorToken


def tokenize(input_string: str) -> list[Token]:
    collected_tokens: list[Token] = []
    operators = ['+', '-', '*', 'x', '/', '(', ')', '^']

    skip_index = 0
    for index in range(len(input_string)):
        if skip_index > 0:
            skip_index -= 1
            continue
        if input_string[index].isspace():
            continue
        if input_string[index].isdecimal():
            literal = ''
            decimal_points_count = 0
            while index + skip_index < len(input_string):
                if input_string[index + skip_index] == '.':
                    literal += input_string[index + skip_index]
                    decimal_points_count += 1
                    if decimal_points_count > 1:
                        raise ValueError(f'Malformed input string. Invalid number ending at index {index+skip_index} found. Too many decimals.')
                elif input_string[index + skip_index].isdecimal():
                    literal += input_string[index + skip_index]
                    # collect if end of literal input
                    if index+skip_index+1 >= len(input_string) or input_string[index+skip_index+1].isspace() or input_string[index+skip_index+1] in operators:
                        try:
                            value = float(literal)
                            collected_tokens.append(LiteralToken(value))
                            break
                        except:
                            raise ValueError(f'Malformed input string. Invalid number ending at index {index+skip_index} found. Malformed decimal number.')
                skip_index += 1
        elif input_string[index] in operators:
            if input_string[index] in ['+', '-'] and (index == 0 or isinstance(collected_tokens[-1], OperatorToken) and collected_tokens[-1].operator != Operator.RPAREN):
                collected_tokens.append(OperatorToken.unary_from_char(input_string[index]))
            else:
                collected_tokens.append(OperatorToken.binary_from_char(input_string[index]))
        else:
            raise ValueError(f'Malformed input string. Unexpected character at index {index} received.')
    return collected_tokens

File: kotlinRpnSolver/src/parse/Tokenizer.kt
package parse

import token.*

fun tokenize(inputString: String) : List<Token> {
    val collectedTokens = ArrayList<Token>()
    val operators = charArrayOf('+', '-', '*', 'x', '/', '^', '(', ')')
    var skipIndex = 0
    for (i in inputString.indices) {
        if (skipIndex > 0) {
            skipIndex--
            continue
        }
        if (inputString[i].isWhitespace()) {
            continue
        }
        if (inputString[i].isDigit()) {
            var literal = ""
            var decimalPointsCount = 0
            while (i + skipIndex < inputString.length) {
                if (inputString[i + skipIndex] == '.'){
                    literal += inputString[i + skipIndex]
                    decimalPointsCount++
                    if (decimalPointsCount > 1) {
                        throw IllegalArgumentException("Malformed input string. Invalid number ending at index ${i+skipIndex} found. Too many decimals.")
                    }
                } else if (inputString[i + skipIndex].isDigit()) {
                    literal += inputString[i + skipIndex]
                    if (i+skipIndex+1 >= inputString.length || inputString[i+skipIndex].isWhitespace() || inputString[i+skipIndex] in operators) {
                        try {
                            val value = literal.toDouble()
                            collectedTokens.add(LiteralToken(value))
                        } catch (ex: NumberFormatException) {
                            throw IllegalArgumentException("Malformed input string. Invalid number ending at index ${i+skipIndex} found. Malformed decimal number.")
                        }
                    }
                }
                skipIndex++
            }
        } else if (inputString[i] in operators) {
            collectedTokens.add(
                if (inputString[i] in setOf('-', '+') && (i == 0 || collectedTokens.last().let { last -> last is OperatorToken && last.operator != Operator.RPAREN }))
                    OperatorToken.unaryFromChar(inputString[i])
                else
                    OperatorToken.binaryFromChar(inputString[i])
            )
        } else {
            throw IllegalArgumentException("Malformed input string. Unexpected character at index $i received.")
        }
    }

    return collectedTokens
}

fun tokenizeFsm(inputString: String) : List<Token> {
    val collectedTokens = ArrayList<Token>();

    val skipCapture = "SKIP"
    val numberCapture = "NUMBER"
    val operatorCapture = "OPERATOR"
    val mismatchCapture = "MISMATCH"

    val tokenRegex = Regex(
        "(?<$numberCapture>\\d+(\\.\\d*)?)|" +
                "(?<$operatorCapture>[+\\-*x/^()])|" +
                "(?<$skipCapture>\\s+)|" +
                "(?<$mismatchCapture>.)"
    )

    tokenRegex.findAll(inputString).forEach { matchResult ->
        when {
            matchResult.groups[skipCapture] != null -> {
                // Skip whitespace
            }
            matchResult.groups[numberCapture] != null -> {
                val value = matchResult.groups[numberCapture]!!.value
                collectedTokens.add(LiteralToken(value.toDouble()))
            }
            matchResult.groups[operatorCapture] != null -> {
                val operatorChar = matchResult.groups[operatorCapture]!!.value[0]
                if (operatorChar in setOf('-', '+') && (collectedTokens.isEmpty() || (collectedTokens.last() is OperatorToken && (collectedTokens.last() as OperatorToken).operator != Operator.RPAREN))) {
                    collectedTokens.add(OperatorToken.unaryFromChar(operatorChar))
                } else {
                    collectedTokens.add(OperatorToken.binaryFromChar(operatorChar))
                }
            }
            matchResult.groups[mismatchCapture] != null -> {
                val value = matchResult.groups[mismatchCapture]!!.value
                collectedTokens.add(InvalidToken(matchResult.range.first, value))
            }
        }
    }
    return collectedTokens
}